cpp/include/cudf/strings/detail/gather.cuh

```diff
@@ -65,67 +72,69 @@ std::unique_ptr<cudf::column> gather(
+                      return d_in_offsets[in_idx + 1] - d_in_offsets[in_idx];
                     });
 
   // check total size is not too large
-  size_t total_bytes = thrust::transform_reduce(
+  size_t const total_bytes = thrust::transform_reduce(
     rmm::exec_policy(stream),
-    d_offsets,
-    d_offsets + output_count,
+    d_out_offsets,
+    d_out_offsets + output_count,
     [] __device__(auto size) { return static_cast<size_t>(size); },
     size_t{0},
     thrust::plus<size_t>{});
   CUDF_EXPECTS(total_bytes < static_cast<std::size_t>(std::numeric_limits<size_type>::max()),
                "total size of output strings is too large for a cudf column");
 
-  // create offsets from sizes
+  // In-place convert output sizes into offsets
   thrust::exclusive_scan(
-    rmm::exec_policy(stream), d_offsets, d_offsets + output_count + 1, d_offsets);
+    rmm::exec_policy(stream), d_out_offsets, d_out_offsets + output_count + 1, d_out_offsets);
 
   // build chars column
-  size_type bytes   = static_cast<size_type>(total_bytes);
-  auto chars_column = create_chars_child_column(output_count, 0, bytes, stream, mr);
-  auto d_chars      = chars_column->mutable_view().template data<char>();
+  size_type const out_chars_bytes = static_cast<size_type>(total_bytes);
+  auto out_chars_column  = create_chars_child_column(output_count, 0, out_chars_bytes, stream, mr);
+  auto const d_out_chars = out_chars_column->mutable_view().template data<char>();
+
   // fill in chars
-  auto gather_chars =
-    [d_strings, begin, strings_count, d_offsets, d_chars] __device__(size_type idx) {
-      auto index = begin[idx];
-      if (NullifyOutOfBounds) {
-        if (is_signed_iterator<MapIterator>() ? ((index < 0) || (index >= strings_count))
-                                              : (index >= strings_count))
-          return;
-      }
-      if (d_strings.is_null(index)) return;
-      string_view d_str = d_strings.element<string_view>(index);
-      memcpy(d_chars + d_offsets[idx], d_str.data(), d_str.size_bytes());
+  cudf::detail::device_span<int32_t const> const d_out_offsets_span(d_out_offsets,
+                                                                    output_count + 1);
+  auto const d_in_chars = (strings_count > 0) ? strings.chars().data<char>() : nullptr;
+  auto gather_chars_fn =
+    [d_out_offsets_span, begin, d_in_offsets, d_in_chars] __device__(size_type out_char_idx) {
+      // find output row index for this output char index
+      auto const next_row_ptr = thrust::upper_bound(
+        thrust::seq, d_out_offsets_span.begin(), d_out_offsets_span.end(), out_char_idx);
+      auto const out_row_idx     = thrust::distance(d_out_offsets_span.begin(), next_row_ptr) - 1;
+      auto const str_char_offset = out_char_idx - d_out_offsets_span[out_row_idx];
+      auto const in_row_idx      = begin[out_row_idx];
+      auto const in_char_offset  = d_in_offsets[in_row_idx] + str_char_offset;
```

Comment:
In the original implementation, the validity of `in_row_idx` is checked if `NullifyOutOfBounds` is `true`. Do we need to check it here as well? If `NullifyOutOfBounds` is `false` and `in_row_idx` is invalid (e.g. -1), what is the expected behavior?

Reply:
The original implementation checked the validity of in_row_idx in all cases, independent of `NullifyOutOfBounds`.  I'm not sure the validity check is necessary for correct semantics, and it's not needed in the char-parallel version.  This gathers the data associated with a row whether that row has valid data or not.  Separately the validity mask will be gathered/generated to decide whether the data that was gathered is valid or not.

Yes, this may mean that in the theoretical worst-case we spend some time gathering bytes that won't matter because those bytes are associated with a null row.  However in practice null string rows have empty data ranges (i.e.: start offset == end offset) so it's faster to avoid examining the validity bit in hopes of avoiding copying extraneous data.

Reply:
I'm not sure we are talking about the same issue. My initial concern is that in the original `gather_chars` kernel, it has
```c++
auto index = begin[idx];
if (NullifyOutOfBounds) {
  if (is_signed_iterator<MapIterator>() ? ((index < 0) || (index >= strings_count))
                                        : (index >= strings_count))
    return;
}
```
so suppose `begin[idx]` returns `-1`, it will be ignored if `NullifyOutOfBounds` is `true`.

In the char-parallel version, this is not checked
```c++
auto const in_row_idx      = begin[out_row_idx];
auto const in_char_offset  = d_in_offsets[in_row_idx] + str_char_offset;
```
so if `begin[out_row_idx]` returns `-1`, the kernel will fail on `d_in_offsets[in_row_idx]`, regardless of `NullifyOutOfBounds`.

On a second thought, I think the current char-parallel implementation is okay. If there is an `idx` such that `begin[idx] = -1`, `d_out_offsets[idx]` should be equal to `d_out_offsets[idx + 1]`, and the binary search performed by `thrust::upper_bound` should guarantee that `out_row_idx` is never equal to `idx`.

Reply:
Ah, yes, sorry I was thrown off when you said "validity".  I thought you meant the column's validity vector. 😅 

> the binary search performed by thrust::upper_bound should guarantee that out_row_idx is never equal to idx

Yep, since we're driving the output character data gather with the output character index, we are always gathering valid characters -- characters that necessarily originate from a valid input row.  Invalid gather map indices were translated to zero-sized output rows when the output offset vector was generated, so there's no output characters that map to an invalid gather map entry.

Reply:
> I'm still confused about what is the expected behavior when NullifyOutOfBounds is false and begin[idx] returns -1

That's invalid input.  I believe the original code would also fail, as it does not check gather map index boundaries unless `NullifyOutOfBounds` is true.  The original code would call `d_strings.is_null(-1)` which would proceed to index the word _before_ the validity vector.  That will either segfault or read some garbage value for validity that if non-zero will then cause it to index before the beginning of the data buffer in a similar fashion.

Reply:
Thanks for the clarifications. I agree the original code would also fail in that case. Just want to confirm that this is indeed the expected behaviors 👍

Reply:
This is intended behavior. If `NullifyOutOfBounds == false` and you pass in an invalid gather map, then you get UB.

Reply:
Yep, that's expected.  The whole point of having the `NullifyOutOfBounds` template parameter is to give the caller the choice of paying the performance cost of bounds-checking the gather map indices or not.  If they choose not to have the checking then the behavior is undefined if they provide invalid gather map indices.